### PR TITLE
Lexer introduce heredoc

### PR DESCRIPTION
**Pull request by kitaisreal**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Introduce syntax for here documents. Example `SELECT $doc$VALUE$doc$`.

### CODE REVIEW

File: src/Parsers/ExpressionElementParsers.cpp [hunk 1]
```diff
@@ -1555,26 +1555,36 @@ bool ParserUnsignedInteger::parseImpl(Pos & pos, ASTPtr & node, Expected & expec
...
+            return false;
+        }
+    }
+    else if (pos->type == TokenType::HereDoc)
     {
-        expected.add(pos, "string literal");
-        return false;
+        std::string_view here_doc(pos->begin, pos->size());
+        size_t heredoc_size = here_doc.find('$', 1) + 1;
+        s = String(pos->begin + heredoc_size, pos->size() - heredoc_size * 2);
```
Comment (alexey-milovidov) [inline src/Parsers/ExpressionElementParsers.cpp:1587]: Add assert pls.

File: tests/queries/0_stateless/01948_heredoc.sql [hunk 2]
```diff
@@ -0,0 +1,4 @@
+SELECT $$VALUE$$;
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/01948_heredoc.sql:1]: Need a test with multiline values.
And also with dollar-something sequences in the middle.

$doc$

$do$
$ doc$
$doc $
$doco$

$doc$;

**Review by alexey-milovidov — CHANGES_REQUESTED** [review]

File: tests/queries/0_stateless/01948_heredoc.sql (hunk 2 above)
Comment (alexey-milovidov) [reply tests/queries/0_stateless/01948_heredoc.sql:1]: Also test with cyrillic/chinese characters - both for separators and for values.
A test with non-UTF8 binary data inside doc will be also nice to have.

Comment (alexey-milovidov) [reply tests/queries/0_stateless/01948_heredoc.sql:1]: Also add .sh test to check how server reacts to unfinished heredocs.
Both unfinished separator and unfinished content. Zero and non-zero size.

File: src/Parsers/ExpressionElementParsers.cpp (hunk 1 above)
Comment (kitaisreal) [reply src/Parsers/ExpressionElementParsers.cpp:1587]: Fixed.

File: tests/queries/0_stateless/01948_heredoc.sql (hunk 2 above)
Comment (kitaisreal) [reply tests/queries/0_stateless/01948_heredoc.sql:2]: Fixed.

File: tests/queries/0_stateless/01948_heredoc.sql [hunk 3]
```diff
@@ -0,0 +1,13 @@
+SELECT $$$$;
+SELECT $$VALUE$$;
+SELECT $doc$$doc$;
+SELECT $doc$VALUE$doc$;
+SELECT $doc$'VALUE'$doc$;
+SELECT $doc$$do$ $ doc$ $doc $ $doco$$doc$;
+SELECT $doc$$do$ $ doc$ $doc $ $doco$$doc$, $doc$$do$ $ doc$ $doc $ $doco$$doc$;
```
Comment (alexey-milovidov) [inline tests/queries/0_stateless/01948_heredoc.sql:7]: No multiline test?
No test with non-UTF-8?

**Comment by sevirov** [comment]
Internal documentation ticket: DOCSUP-12136

**Comment by alexey-milovidov** [comment]
@kitaisreal Still waiting for test with multiline and non-UTF-8 entries.
Also `normalizeQuery`, `normalizedQueryHash` functions have to be adjusted.
Please also add a test for these functions.

**Pull request merged**